### PR TITLE
Quick xplain proposal

### PR DESCRIPTION
**Pull request by schmittjoaopedro**
I'm proposing this PR @cprudhom so that you will be able to better evaluate what we were talking about in gitter.

Changes proposed in this PR:
- Add a QuickXPlain algorithm to find the minimum conflicting set of constraints (ISSUE #509)
- Add support to disable constraints from execution (this feature is also required for QuickXPlain)
 
@chocoteam/core-developer

### CODE REVIEW

File: solver/src/main/java/org/chocosolver/solver/Solver.java [hunk 1]
```diff
@@ -626,6 +626,15 @@ public void propagate() throws ContradictionException {
         }
     }
 
+    /**
+     * Return the minimum conflicting set from a conflicting set that is causing contradiction.
+     * @param conflictingSet
+     * @return minimumConflictingSet
+     */
+    public List<Constraint> findMinimumConflictingSet(List<Constraint> conflictingSet) {
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/Solver.java:634]: I think this is a good place to add assertions and eventually throw exception if looking for the MCS is called in inappropriate way (during solving for instance). The javadoc should also exhibit these assertions.

File: solver/src/main/java/org/chocosolver/solver/constraints/Constraint.java [hunk 2]
```diff
@@ -84,6 +86,11 @@
      */
     private String name;
 
+    /**
+     * If a constraint is enabled to the propagation engine.
+     */
+    private boolean enabled = Boolean.TRUE;
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/Constraint.java:92]: Shouldn't it be `true`? What is the benefit of using Boolean.TRUE?

File: solver/src/main/java/org/chocosolver/solver/constraints/Constraint.java [hunk 3]
```diff
@@ -391,4 +398,37 @@ public static Constraint merge(String name, Constraint... toMerge) {
         }
         return new Constraint(name, props.toArray(new Propagator[0]));
     }
+
+    public boolean isEnabled() {
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/Constraint.java:402]: Could you add some javadoc?

File: solver/src/main/java/org/chocosolver/solver/constraints/Propagator.java [hunk 4]
```diff
@@ -100,6 +100,11 @@
      */
     private static final short PASSIVE = 3;
 
+    /**
+     * Ignore propagation during execution.
+     */
+    private boolean enabled = Boolean.TRUE;
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/Propagator.java:106]: Same question as previous: I would have expected `true`

**Review by cprudhom — CHANGES_REQUESTED** [review]

File: solver/src/main/java/org/chocosolver/solver/Solver.java (hunk 1 above)
Comment (schmittjoaopedro) [reply solver/src/main/java/org/chocosolver/solver/Solver.java:636]: Sure, I add the check for solving and created a unitary test to thist.

File: solver/src/main/java/org/chocosolver/solver/constraints/Constraint.java (hunk 2 above)
Comment (schmittjoaopedro) [reply solver/src/main/java/org/chocosolver/solver/constraints/Constraint.java]: Sorry, it's a costume that I have.
I checked on the internet and to use the primite `true` only requires a shift in registers (that's more efficient than initialize a static class). In fixed to use primitive `true`.

File: solver/src/main/java/org/chocosolver/solver/constraints/Constraint.java (hunk 3 above)
Comment (schmittjoaopedro) [reply solver/src/main/java/org/chocosolver/solver/constraints/Constraint.java:411]: Yes, I added on the last commit.

File: solver/src/main/java/org/chocosolver/solver/constraints/Propagator.java (hunk 4 above)
Comment (schmittjoaopedro) [reply solver/src/main/java/org/chocosolver/solver/constraints/Propagator.java]: Fixed.

**Review by cprudhom — APPROVED** [review]

File: solver/src/main/java/org/chocosolver/solver/Solver.java [hunk 5]
```diff
@@ -793,6 +807,20 @@ public boolean moveBackward(){
...
+     * Solving is executing if the search state is different from NEW, that is,
+     * if it has started to branch decisions.
+     * A double check for execution is done looking if the environment trailing
+     * has started as well.
+     *
+     * @return isSolving if the solver is executing searching or branching
+     */
+    public boolean isSolving() {
+        boolean isSearching = getSearchState() != SearchState.NEW;
+        boolean isTrailing = getEnvironment().getWorldIndex() > 0;
```
[the remark this answers is not in the document]
Comment (schmittjoaopedro) [reply solver/src/main/java/org/chocosolver/solver/Solver.java]: Fixed.

**Comment by schmittjoaopedro** [comment]
I've taken the freedom to add the QuickXPlain feature item in `CHANGES.md`

**Review by cprudhom — APPROVED** [review]

**Pull request merged**